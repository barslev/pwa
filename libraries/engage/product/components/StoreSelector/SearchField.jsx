import React, { Fragment, useState } from 'react';
import PropTypes from 'prop-types';
import { i18n, withCurrentProduct } from '@shopgate/engage/core';
import { ProgressBar, MagnifierIcon, LocatorIcon } from '@shopgate/engage/components';
import connect from './SearchField.connector';
import {
  container, search, input, icon, progressBar,
} from './SearchField.style';

/**
 * The SearchField component.
 * @param {Object} props The component props.
 * @returns {JSX}
 */
function SearchField(props) {
  const {
    productId,
    loading,
    getLocationsByPostalCode,
    getLocationsByGeolocation,
  } = props;

  const [query, setQuery] = useState('');

  /**
   * Updates the query state of the component when the input changes.
   * @param {Object} e A React event object.
   */
  const handleOnChange = (e) => {
    setQuery(e.target.value);
  };

  /**
   * Queries locations by postal codes.
   * @param {Object} e A React event object.
   */
  const handleOnKeyDown = (e) => {
    if (e.keyCode === 13) {
      getLocationsByPostalCode(productId, query);
    }
  };

  /**
   * Queries locations by geo coordinates.
   */
  const handleLocate = () => {
    getLocationsByGeolocation(productId);
  };

  return (
    <Fragment>
      <div className={container}>
        <div className={search}>
          <span className={icon} aria-hidden>
            <MagnifierIcon />
          </span>
          <input
            className={input}
            value={query}
            onChange={handleOnChange}
            onKeyDown={handleOnKeyDown}
            disabled={loading}
            type="search"
            autoComplete="off"
            autoCorrect="off"
            placeholder={i18n.text('product.zip_search')}
          />
          <button onClick={handleLocate} type="button" className={icon}>
            <LocatorIcon />
          </button>
        </div>
      </div>
      <div className={progressBar}>
        <ProgressBar isVisible={loading} />
      </div>
    </Fragment>
  );
}

SearchField.propTypes = {
  getLocationsByGeolocation: PropTypes.func.isRequired,
  getLocationsByPostalCode: PropTypes.func.isRequired,
  loading: PropTypes.bool,
  productId: PropTypes.string,
};

SearchField.defaultProps = {
  productId: null,
  loading: false,
};

export const SearchFieldUnwrapped = SearchField;

export default withCurrentProduct(connect(SearchField));
