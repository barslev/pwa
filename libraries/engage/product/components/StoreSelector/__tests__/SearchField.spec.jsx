import React from 'react';
import { shallow, mount } from 'enzyme';
import { SearchFieldUnwrapped as SearchField } from '../SearchField';

jest.mock('@shopgate/engage/components', () => ({
  LocatorIcon: () => '',
  MagnifierIcon: () => '',
  ProgressBar: () => '',
}));

describe('<SearchField />', () => {
  const getLocationsByGeolocation = jest.fn();
  const getLocationsByPostalCode = jest.fn();

  const productId = 'ABC123';
  const defaultProps = {
    productId,
    getLocationsByGeolocation,
    getLocationsByPostalCode,
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render as expected', () => {
    const wrapper = shallow((
      <SearchField {...defaultProps} />
    ));
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('ProgressBar').prop('isVisible')).toEqual(false);
    expect(wrapper.find('input').prop('disabled')).toEqual(false);
  });

  it.only('should react on updates of the loading prop', () => {
    const wrapper = shallow((
      <SearchField {...defaultProps} />
    ));

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('ProgressBar').prop('isVisible')).toEqual(false);
    expect(wrapper.find('input').prop('disabled')).toEqual(false);

    wrapper.setProps({ loading: true });
    expect(wrapper.find('ProgressBar').prop('isVisible')).toEqual(true);
    expect(wrapper.find('input').prop('disabled')).toEqual(true);

    wrapper.setProps({ loading: false });
    expect(wrapper.find('ProgressBar').prop('isVisible')).toEqual(false);
    expect(wrapper.find('input').prop('disabled')).toEqual(false);
  });

  it('should call getLocationsByPostalCode after filling the input and pressing enter', () => {
    const wrapper = mount((
      <SearchField {...defaultProps} />
    ));
    const mockValue = 'ACME';
    const input = wrapper.find('input');
    input.simulate('change', { target: { value: mockValue } });
    input.simulate('keyDown', { keyCode: 13 });
    expect(wrapper).toMatchSnapshot();
    expect(getLocationsByPostalCode).toHaveBeenCalledTimes(1);
    expect(getLocationsByPostalCode).toHaveBeenCalledWith(productId, mockValue);
  });

  it('should call getLocationsByGeolocation after clicking the locate button', () => {
    const wrapper = shallow((
      <SearchField {...defaultProps} />
    ));
    const button = wrapper.find('button');
    button.simulate('click');
    expect(wrapper).toMatchSnapshot();
    expect(getLocationsByGeolocation).toHaveBeenCalledTimes(1);
    expect(getLocationsByGeolocation).toHaveBeenCalledWith(productId);
  });
});
