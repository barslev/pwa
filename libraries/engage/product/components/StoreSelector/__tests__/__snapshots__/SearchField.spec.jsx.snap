// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchField /> should call getLocationsByGeolocation after clicking the locate button 1`] = `
<Fragment>
  <div
    className={
      Object {
        "data-css-z3aef1": "",
      }
    }
  >
    <div
      className={
        Object {
          "data-css-bpbkyq": "",
        }
      }
    >
      <span
        aria-hidden={true}
        className={
          Object {
            "data-css-gy1p1e": "",
          }
        }
      >
        <MagnifierIcon />
      </span>
      <input
        autoComplete="off"
        autoCorrect="off"
        className={
          Object {
            "data-css-2h0g54": "",
          }
        }
        disabled={false}
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="product.zip_search"
        type="search"
        value=""
      />
      <button
        className={
          Object {
            "data-css-gy1p1e": "",
          }
        }
        onClick={[Function]}
        type="button"
      >
        <LocatorIcon />
      </button>
    </div>
  </div>
  <div
    className={
      Object {
        "data-css-13azwyo": "",
      }
    }
  >
    <ProgressBar
      isVisible={false}
    />
  </div>
</Fragment>
`;

exports[`<SearchField /> should call getLocationsByPostalCode after filling the input and pressing enter 1`] = `
<SearchField
  getLocationsByGeolocation={[MockFunction]}
  getLocationsByPostalCode={
    [MockFunction] {
      "calls": Array [
        Array [
          "ABC123",
          "ACME",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  loading={false}
  productId="ABC123"
>
  <div
    className={
      Object {
        "data-css-z3aef1": "",
      }
    }
  >
    <div
      className={
        Object {
          "data-css-bpbkyq": "",
        }
      }
    >
      <span
        aria-hidden={true}
        className={
          Object {
            "data-css-gy1p1e": "",
          }
        }
      >
        <MagnifierIcon />
      </span>
      <input
        autoComplete="off"
        autoCorrect="off"
        className={
          Object {
            "data-css-2h0g54": "",
          }
        }
        disabled={false}
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="product.zip_search"
        type="search"
        value="ACME"
      />
      <button
        className={
          Object {
            "data-css-gy1p1e": "",
          }
        }
        onClick={[Function]}
        type="button"
      >
        <LocatorIcon />
      </button>
    </div>
  </div>
  <div
    className={
      Object {
        "data-css-13azwyo": "",
      }
    }
  >
    <ProgressBar
      isVisible={false}
    />
  </div>
</SearchField>
`;

exports[`<SearchField /> should react on updates of the loading prop 1`] = `
<Fragment>
  <div
    className={
      Object {
        "data-css-z3aef1": "",
      }
    }
  >
    <div
      className={
        Object {
          "data-css-bpbkyq": "",
        }
      }
    >
      <span
        aria-hidden={true}
        className={
          Object {
            "data-css-gy1p1e": "",
          }
        }
      >
        <MagnifierIcon />
      </span>
      <input
        autoComplete="off"
        autoCorrect="off"
        className={
          Object {
            "data-css-2h0g54": "",
          }
        }
        disabled={false}
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="product.zip_search"
        type="search"
        value=""
      />
      <button
        className={
          Object {
            "data-css-gy1p1e": "",
          }
        }
        onClick={[Function]}
        type="button"
      >
        <LocatorIcon />
      </button>
    </div>
  </div>
  <div
    className={
      Object {
        "data-css-13azwyo": "",
      }
    }
  >
    <ProgressBar
      isVisible={false}
    />
  </div>
</Fragment>
`;

exports[`<SearchField /> should render as expected 1`] = `
<Fragment>
  <div
    className={
      Object {
        "data-css-z3aef1": "",
      }
    }
  >
    <div
      className={
        Object {
          "data-css-bpbkyq": "",
        }
      }
    >
      <span
        aria-hidden={true}
        className={
          Object {
            "data-css-gy1p1e": "",
          }
        }
      >
        <MagnifierIcon />
      </span>
      <input
        autoComplete="off"
        autoCorrect="off"
        className={
          Object {
            "data-css-2h0g54": "",
          }
        }
        disabled={false}
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="product.zip_search"
        type="search"
        value=""
      />
      <button
        className={
          Object {
            "data-css-gy1p1e": "",
          }
        }
        onClick={[Function]}
        type="button"
      >
        <LocatorIcon />
      </button>
    </div>
  </div>
  <div
    className={
      Object {
        "data-css-13azwyo": "",
      }
    }
  >
    <ProgressBar
      isVisible={false}
    />
  </div>
</Fragment>
`;
